[Docs] Improve druid.coordinator.kill.on description (#17538)
apache/druid: docs/configuration/index.md

|`druid.coordinator.load.timeout`|The timeout duration for when the Coordinator assigns a segment to a Historical service.|`PT15M`|
|`druid.coordinator.kill.pendingSegments.on`|Boolean flag for whether or not the Coordinator clean up old entries in the `pendingSegments` table of metadata store. If set to true, Coordinator will check the created time of most recently complete task. If it doesn't exist, it finds the created time of the earliest running/pending/waiting tasks. Once the created time is found, then for all datasources not in the `killPendingSegmentsSkipList` (see [Dynamic configuration](#dynamic-configuration)), Coordinator will ask the Overlord to clean up the entries 1 day or more older than the found created time in the `pendingSegments` table. This will be done periodically based on `druid.coordinator.period.indexingPeriod` specified.|true|

> |`druid.coordinator.kill.on`|Boolean flag for whether to submit a kill task for unused segments, permanently deleting them from the metadata store and deep storage. When `druid.coordinator.kill.on` is true: <ul><li>Segments become eligible for permanent deletion once their data intervals are older than `druid.coordinator.kill.durationToRetain`  relative to the current time. If a segment's data interval exceeds this threshold at the time it is marked unused, it becomes eligible for immediate deletion.</li><li>If `killDataSourceWhitelist` specifies data sources, the Coordinator issues kill tasks only for those data sources based on the specified period.</li><li>If `killDataSourceWhitelist` is empty, the Coordinator issues kill tasks for all data sources.</li></ul>|false|

Reviewer: ```suggestion
|`druid.coordinator.kill.on`|Boolean flag to submit a kill task for unused segments and delete them permanently from the metadata store and deep storage. When `druid.coordinator.kill.on` is true: <ul><li>Segments become eligible for permanent deletion once their data intervals are older than `druid.coordinator.kill.durationToRetain`  relative to the current time. If a segment's data interval exceeds this threshold at the time it is marked unused, it becomes eligible for immediate deletion.</li><li>If `killDataSourceWhitelist` specifies data sources, the Coordinator issues kill tasks only for those data sources based on the specified period.</li><li>If `killDataSourceWhitelist` is empty, the Coordinator issues kill tasks for all data sources.</li></ul>|false|
```

> |`druid.coordinator.kill.on`|Boolean flag to enable the Coordinator to submit a kill task for unused segments and delete them permanently from the metadata store and deep storage. When `druid.coordinator.kill.on` is true: <ul><li>Segments become eligible for permanent deletion once their data intervals are older than `druid.coordinator.kill.durationToRetain`  relative to the current time. If a segment's data interval exceeds this threshold at the time it is marked unused, it becomes eligible for immediate deletion.</li><li>If `killDataSourceWhitelist` specifies data sources, the Coordinator issues kill tasks only for those data sources based on the specified period.</li><li>If `killDataSourceWhitelist` is empty, the Coordinator issues kill tasks for all data sources.</li></ul>|false|

Reviewer 2: Simplified the language a bit.
Removed the part about becoming eligible immediately since we now have the concept of a `bufferPeriod`.

```suggestion
|`druid.coordinator.kill.on`|Boolean flag to enable the Coordinator to submit a kill task for unused segments and delete them permanently from the metadata store and deep storage. When `druid.coordinator.kill.on` is true: <ul><li>Segments become eligible for permanent deletion once their data intervals become older than `druid.coordinator.kill.durationToRetain` relative to the current time.</li><li>If `killDataSourceWhitelist` specifies data sources, the Coordinator issues kill tasks only for those data sources.</li><li>If `killDataSourceWhitelist` is empty, the Coordinator issues kill tasks for all data sources.</li></ul>|false|
```

Reviewer 3: Could we also add a brief note on the impact of `bufferPeriod` when determining the intervals or segments to kill (similar to what we have here for `durationToRetain`)? Essentially only unused segments in the metadata store that have a `used_status_last_updated` time earlier than the configured `bufferPeriod` are eligible for deletion.

Reviewer 2: I wonder if we need that since it would already be described in the entry for `bufferPeriod` itself. For that matter, I would suggest even getting rid of the part about `durationToRetain` and `killDataSourceWhitelist`.

Author: I agree with removing references to `durationToRetain` and `killDataSourceWhiteList` from the `druid.coordinator.kill.on` description. @kfaraz @abhishekrb19 Do you want to move forward with this change?

|`druid.coordinator.kill.pendingSegments.on`|Boolean flag for whether or not the Coordinator clean up old entries in the `pendingSegments` table of metadata store. If set to true, Coordinator will check the created time of most recently complete task. If it doesn't exist, it finds the created time of the earliest running/pending/waiting tasks. Once the created time is found, then for all datasources not in the `killPendingSegmentsSkipList` (see [Dynamic configuration](#dynamic-configuration)), Coordinator will ask the Overlord to clean up the entries 1 day or more older than the found created time in the `pendingSegments` table. This will be done periodically based on `druid.coordinator.period.indexingPeriod` specified.|true|
|`druid.coordinator.kill.on`|Boolean flag to enable the Coordinator to submit a kill task for unused segments and delete them permanently from the metadata store and deep storage.|false|
|`druid.coordinator.kill.period`| The frequency of sending kill tasks to the indexing service. The value must be greater than or equal to `druid.coordinator.period.indexingPeriod`. Only applies if kill is turned on.|Same as `druid.coordinator.period.indexingPeriod`|

> |`druid.coordinator.kill.durationToRetain`|Period to retain unused segments in ISO 8601 duration format. When `druid.coordinator.kill.on` is true, segments become eligible for permanent deletion once their data intervals become older than `durationToRetain` relative to the current time. For example, if `durationToRetain` is set to `P90D`, a segment from a time interval 90 days in the past becomes eligible for permanent deletion immediately after being marked unused. If you set `durationToRetain` to a negative ISO 8601 period, `now - durationToRetain` falls in the future.|`P90D`|

Reviewer 2: Unfortunately, the config names are a little confusing.
The `bufferPeriod` should really be called the `durationToRetain` since the buffer period represents the duration for which we retain the segments even after they have been marked as unused.
The `durationToRetain` however is the eligibility criteria for segments to be deleted.
I have tried to update the language below to reflect that.

```suggestion
|`druid.coordinator.kill.durationToRetain`|Duration (in ISO 8601 format) relative to the current time identifying data interval of segments which should be retained. When `druid.coordinator.kill.on` is true, any segment with a data interval ending before `now - durationToRetain` is eligible for permanent deletion. For example, if `durationToRetain` is set to `P90D`, an unused segment with a time interval ending 90 days in the past is eligible for deletion. If `durationToRetain` is set to a negative ISO 8601 period, segments with future intervals ending before `now - durationToRetain` would be eligible for deletion too.|`P90D`|
```

Author: Updated

|`replicantLifetime`|The maximum number of Coordinator runs for which a segment can wait in the load queue of a Historical before Druid raises an alert.|15|
|`replicationThrottleLimit`|The maximum number of segment replicas that can be assigned to a historical tier in a single Coordinator run. This property prevents Historical services from becoming overwhelmed when loading extra replicas of segments that are already available in the cluster.|500|
|`balancerComputeThreads`|Thread pool size for computing moving cost of segments during segment balancing. Consider increasing this if you have a lot of segments and moving segments begins to stall.|`num_cores` / 2|

> |`killDataSourceWhitelist`|List of specific data sources for which kill tasks are sent if property `druid.coordinator.kill.on` is true. Can be a list of comma-separated data source names or a JSON array. If `killDataSourceWhitelist` is empty, the Coordinator issues kill tasks for all data sources.|none|

Reviewer 2: ```suggestion
|`killDataSourceWhitelist`|List of specific data sources for which kill tasks can be issued if `druid.coordinator.kill.on` is true. It can be a comma-separated list of data source names or a JSON array. If `killDataSourceWhitelist` is empty, the Coordinator issues kill tasks for all data sources.|none|
```

Author: Updated

> |`killDataSourceWhitelist`|List of specific data sources for which kill tasks are sent if property `druid.coordinator.kill.on` is true. If `killDataSourceWhitelist` is empty, the Coordinator issues kill tasks for all data sources. This can be a list of comma-separated data source names or a JSON array.|none|

Author: ```suggestion
|`killDataSourceWhitelist`|List of specific data sources for which kill tasks are sent if property `druid.coordinator.kill.on` is true. Can be a list of comma-separated data source names or a JSON array. If `killDataSourceWhitelist` is empty, the Coordinator issues kill tasks for all data sources.|none|
```
